image/manifest.go

```diff
@@ -240,3 +240,34 @@ loop:
 	}
 	return nil
 }
+
+// ValidateManifestMediaType validate the manifest schema media-type
+// and its descendants fields media type, such as config and layers,
+// to make sure they match to spec definition, or returns an error if
+// the validation failed.
+func ValidateManifestMediaType(r io.ReadCloser) error {
+	header := v1.Manifest{}
+	defer r.Close()
+
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return errors.Wrapf(err, "error reading the io stream")
+	}
+
+	err = json.NewDecoder(bytes.NewReader(buf)).Decode(&header)
+	if err != nil {
+		return errors.Wrap(err, "manifest format mismatch")
+	}
```

Comment:
Hmm, with 739f3ad → a574b2f you dropped the early `defer r.Close()`.  I think you want to keep that so the reader is closed on early errors.

And there's no reason to read the file in and then create a `bytes.NewReader` around it.  Just call `json.NewDecoder(r)…`.